packages/gatsby/src/commands/serve.js

```diff
@@ -40,6 +41,29 @@ const readMatchPaths = async program => {
   return JSON.parse(rawJSON)
 }
 
+/**
+ * Adds trailing slash to request.url, in order to deceive static middleware
+ * into getting a directory request, thus preventing redirect.
+ * This is an express middleware to be used before static middleware
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ */
+const directoryRedirectPrevention = (req, res, next) => {
+  // url is home page
+  if (req.url === `/`) return next()
+  // url has trailing slash, no need to intervene
+  if (req.url.slice(-1) === `/`) return next()
```

Comment:
The second check subsumes the first because `'x'.slice(-1) === 'x'` :)

Reply:
That's true, I wanted to differentiate the two cases, easier to understand.
Falls into "write code for humans" practice. The fact that you noticed, proves my point.

Reply:
I get what you're saying but I still think I'd prefer to drop the redundant check and updating the comment to reflect the two cases being caught here.